yksgi/src/lib.rs

```diff
@@ -313,8 +316,68 @@ impl SGInterp {
             // FIXME: When we see the control point we are done and can just return.
             return true;
         } else {
-            // FIXME: Properly implement calls.
-            todo!()
+            // We are going to do a native call via libffi, starting with looking up the address of
+            // the callee in the virtual address space.
+            //
+            // OPT: This could benefit from caching, and we may already know the address of the
+            // callee if we inlined it when preparing traces:
+            // https://github.com/ykjit/yk/issues/544
+            assert!(!LLVMIsAFunction(func).is_null());
+            let fptr = dlsym(ptr::null_mut(), name.as_ptr() as *const i8);
+            if fptr == ptr::null_mut() {
+                todo!("couldn't find symbol: {}", name);
+            }
+
+            if LLVMIsFunctionVarArg(LLVMGetElementType(LLVMTypeOf(func))) != 0 {
+                todo!("calling a varargs function");
+            }
+
+            // Now build the calling interface and collect the values of the callee's arguments.
+            //
+            // It may be tempting to use `LLVMGetArgOperand()` to extract callee operands from our
+            // `CallInst`, but at the time of writing this wraps the C++ method
+            // `FuncletPadInst::getArgOperand()` which is not what we want.
+            //
+            // `CallBase::getArgOperand()` isn't currently exposed by the LLVM C API, so we instead
+            // use the implicit knowledge that for a `CallInst` the first `N` operands are the
+            // operands of the function being called (the remaining operand is the callee).
+            //
+            // Despite this, `LLVMGetNumArgOperands()` *is* generic and does the right thing for
+            // any callable function-like thing, so we can still use that for the loop bounds.
+            let mut builder = FFIBuilder::new();
+            let mut arg_vals = Vec::new();
+            for i in 0..LLVMGetNumArgOperands(call_inst) {
+                let arg = Value::new(LLVMGetOperand(call_inst, i.try_into().unwrap()));
```

Comment:
if you wrap call_inst inside `Value`, e.g. `Value::new(call_inst)`, then you can use `get_operand(idx)`. I would probably do that in line 288. Oh, now that I look at it, it is already wrapped `self.pc`. So you could just leave it in `self.pc`.

Reply:
I reviewed this part of code and overhauled some of the wrapper usage. See d0327ba5d.